presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -586,10 +577,12 @@ private PlanWithProperties planTableScan(TableScanNode node, Expression predicat
                     deterministicPredicate,
                     DomainTranslator.toPredicate(node.getCurrentConstraint().transform(assignments::get)));
 
+            ConstraintEvaluator evaluator = new ConstraintEvaluator(session, metadata, parser, types, constraint);
+
             // Layouts will be returned in order of the connector's preference
             List<TableLayoutResult> layouts = metadata.getLayouts(
                     session, node.getTable(),
-                    new Constraint<>(simplifiedConstraint, bindings -> !shouldPrune(constraint, node.getAssignments(), bindings, context.getCorrelations())),
+                    new Constraint<>(simplifiedConstraint, bindings -> !evaluator.shouldPrune(node.getAssignments(), bindings, context.getCorrelations())),
```

Comment:
`node.getAssignments()` and `context.getCorrelations()` are constant, I think they should be passed to constructor. Then you could filter out some conjuncts, so there will be lower cost of evaluating this statement per each partition.